From 905de87eb264ae90e5f8fe0de5b018247c0a20c0 Mon Sep 17 00:00:00 2001
From: Aram So <aram.so@lge.com>
Date: Wed, 13 May 2015 11:20:47 +0900
Subject: [PATCH] Allow fixed animation step in QSGGuiThreadRenderLoop

In Qt 5.4 the fixed animation step becomes available only for
QSGWindowsRenderLoop and QSGThreadedRenderLoop. However in webOS,
we are still using QSGGuiThreadRenderLoop and some wayland clients need
this option for the animation smoothness.
This patch allows to enable the fixed animation step in
QSGGuiThreadRenderLoop if QSG_FIXED_ANIMATION_STEP is set.

Related upstream patch: https://codereview.qt-project.org/#/c/94737/

Upstream-Status: Pending
---
 src/quick/scenegraph/qsgrenderloop.cpp | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/src/quick/scenegraph/qsgrenderloop.cpp b/src/quick/scenegraph/qsgrenderloop.cpp
index c4111a9..caba690 100644
--- a/src/quick/scenegraph/qsgrenderloop.cpp
+++ b/src/quick/scenegraph/qsgrenderloop.cpp
@@ -126,6 +126,8 @@ public:
     QSGGuiThreadRenderLoop();
     ~QSGGuiThreadRenderLoop();
 
+    static bool useConsistentTiming();
+
     void show(QQuickWindow *window);
     void hide(QQuickWindow *window);
 
@@ -267,6 +269,11 @@ QSGGuiThreadRenderLoop::QSGGuiThreadRenderLoop()
     : gl(0)
     , eventPending(false)
 {
+    if (useConsistentTiming()) {
+        QUnifiedTimer::instance(true)->setConsistentTiming(true);
+        qCDebug(QSG_LOG_INFO, "using fixed animation steps");
+    }
+
     sg = QSGContext::createDefaultContext();
     rc = sg->createRenderContext();
 
@@ -280,6 +287,15 @@ QSGGuiThreadRenderLoop::~QSGGuiThreadRenderLoop()
     delete sg;
 }
 
+bool QSGGuiThreadRenderLoop::useConsistentTiming()
+{
+    // Enable fixed animation steps if QSG_FIXED_ANIMATION_STEP is set
+    QByteArray fixedAnimationSteps = qgetenv("QSG_FIXED_ANIMATION_STEP");
+    if (fixedAnimationSteps.length())
+        return true;
+    return false;
+}
+
 void QSGGuiThreadRenderLoop::show(QQuickWindow *window)
 {
     WindowData data;
